fpga_interchange/arch.cc

```diff
@@ -573,4 +693,95 @@ const std::vector<std::string> Arch::availablePlacers = {"sa",
 const std::string Arch::defaultRouter = "router2";
 const std::vector<std::string> Arch::availableRouters = {"router1", "router2"};
 
+void Arch::map_cell_pins(CellInfo *cell, int32_t mapping) const
+{
+    cell->cell_mapping = mapping;
+    cell->pins.clear();
+    cell->cell_bel_pins.clear();
+
+    const CellBelMapPOD &cell_pin_map = chip_info->cell_map->cell_bel_map[mapping];
+
+    log_info("Mapping cell %s\n", cell->name.c_str(this));
+    for (const auto &pin_map : cell_pin_map.common_pins) {
+        IdString cell_pin(pin_map.cell_pin);
+        IdString bel_pin(pin_map.bel_pin);
+
+        if(cell_pin.str(this) == "GND") {
+            // FIXME: Tie this pin to the GND net
+            continue;
+        }
+        if(cell_pin.str(this) == "VCC") {
+            // FIXME: Tie this pin to the VCC net
+            continue;
+        }
+
+        auto &bel_pins = cell->cell_bel_pins[cell_pin];
+        log_info("%s => %s\n", cell_pin.c_str(this), bel_pin.c_str(this));
+        cell->pins[cell_pin] = bel_pin;
+        bel_pins.push_back(bel_pin);
+
+        // FIXME: Only have support for 1-to-1 pin in the router right now.
+        NPNR_ASSERT(bel_pins.size() == 1);
+    }
+
+    for (const auto &parameter_pin_map : cell_pin_map.parameter_pins) {
+        IdString param_key(parameter_pin_map.key);
+        std::string param_value = IdString(parameter_pin_map.value).c_str(this);
+
+        auto iter = cell->params.find(param_key);
+        if (iter == cell->params.end()) {
+            continue;
+        }
+
+        for (const auto &pin_map : cell_pin_map.common_pins) {
+            if (param_value == iter->second.as_string()) {
+                IdString cell_pin(pin_map.cell_pin);
+                IdString bel_pin(pin_map.bel_pin);
+
+                cell->pins[cell_pin] = bel_pin;
+                auto &bel_pins = cell->cell_bel_pins[cell_pin];
+                bel_pins.push_back(bel_pin);
+
+                // FIXME: Only have support for 1-to-1 pin in the router right
+                // now.
+                NPNR_ASSERT(bel_pins.size() == 1);
+            }
+        }
+    }
+}
+
+void Arch::write_physical_netlist(const std::string &filename) const
+{
+    FpgaInterchange::write_physical_netlist(getCtx(), filename);
+}
+
+void Arch::read_logical_netlist(const std::string &filename)
+{
+    FpgaInterchange::read_logical_netlist(getCtx(), filename);
+}
+
+std::string Arch::get_part() const {
+    // FIXME: Need a map between device / package / speed grade and part.
+    return chip_info->name.get() + args.package + "-1";
+}
+
+void Arch::map_port_pins(BelId bel, CellInfo *cell) const {
+    IdStringRange pins = getBelPins(bel);
+    NPNR_ASSERT(pins.begin() != pins.end());
```

Comment:
I think I've seen a few places where you are getting exactly one item from an iterator. It would be neater to have a generic function (e.g. in `util.h`) with the various asserts for this purpose

Reply:
Sure.

Reply:
Done!